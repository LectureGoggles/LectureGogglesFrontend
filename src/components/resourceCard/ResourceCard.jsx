import React, { useContext, useState, useCallback, useReducer, useEffect } from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import axios from 'axios';
import { Formik } from 'formik';

import AuthContext from '../../contexts/AuthContext';
import {
  UnstyledButton,
  CardContainerStyle,
  TitleStyle,
  ItemStyle,
  DescriptionStyle,
  AvatarStyle,
  PreviewStyle,
  PreviewA,
  PreviewLink,
  LinkStyle,
  BottomContainer,
  TextAreaStyle
} from '../__styles__/styles';
import ResourceCardDropDown from './DropDown';
import Modal from './reportModal/Modal';
import GenericButton from '../button/button';

const urlToUse = process.env.NODE_ENV === 'development' ? '' : 'https://api.lecturegoggles.io';

const ResourceCard = ({
  title,
  subject,
  topic,
  author,
  authorImg,
  previewImg,
  points,
  timeStamp,
  description,
  url,
  id,
  vote
}) => {
  const { signedInAs, userData } = useContext(AuthContext);

  const [pointState, setPoints] = useState(() => {
    if (typeof vote !== 'undefined') {
      if (vote.vote_choice === 1) {
        return points - 1;
      }
      if (vote.vote_choice === -1) {
        return points + 1;
      }
    }
    return points;
  });

  const [dateToDisplay] = useState(() => {
    const date = new Date(timeStamp);
    return date.toDateString();
  });

  const [isShowingDropDown, setShowingDropDown] = useState(false);

  const [isShowingReportModal, setShowingReportModal] = useReducer((state, action) => {
    setShowingDropDown(false);
    return action;
  }, false);

  const [isShowingImageModal, setShowingImageModal] = useReducer((state, action) => {
    setShowingDropDown(false);
    return action;
  }, false);

  useEffect(() => {
    document.addEventListener('keydown', event => {
      if (event.key === 'Escape') {
        setShowingReportModal(false);
        setShowingImageModal(false);
      }
    });
  }, []);

  function reportPost(values, actions) {
    const token = localStorage.getItem('token');
    if (values.description !== '') {
      axios
        .post(
          `${urlToUse}/v1/report/createReport/${id}/`,
          { description: values.description },
          { headers: { Authorization: `Bearer ${token}` } }
        )
        .then(() => actions.resetForm() && actions.setSubmitting(false));
    }
  }

  function addImageToPost(values, actions) {
    const token = localStorage.getItem('token');
    if (values.description !== '') {
      axios
        .post(
          `${urlToUse}/v1/post/setPostmageOn/${id}/`,
          { url: values.url },
          { headers: { Authorization: `Bearer ${token}` } }
        )
        .then(() => actions.resetForm() && actions.setSubmitting(false));
    }
  }

  const createReducer = () => (state, action) => {
    const token = localStorage.getItem('token');
    if (state === 'upvote') {
      if (action === 'upvote') {
        setPoints(pointState);
        axios.post(`${urlToUse}/v1/vote/upvotePost/${id}/`, {}, { headers: { Authorization: `Bearer ${token}` } });
        return 'none';
      }
    }
    if (state === 'downvote') {
      if (action === 'downvote') {
        setPoints(pointState);
        axios.post(`${urlToUse}/v1/vote/downvotePost/${id}/`, {}, { headers: { Authorization: `Bearer ${token}` } });
        return 'none';
      }
    }
    if (action === 'upvote') {
      setPoints(pointState + 1);
      axios.post(`${urlToUse}/v1/vote/upvotePost/${id}/`, {}, { headers: { Authorization: `Bearer ${token}` } });
    }
    if (action === 'downvote') {
      setPoints(pointState - 1);
      axios.post(`${urlToUse}/v1/vote/downvotePost/${id}/`, {}, { headers: { Authorization: `Bearer ${token}` } });
    }
    return action;
  };

  const memoizedCallback = useCallback(createReducer('none'), []);

  const [voteState, dispatch] = useReducer(memoizedCallback, 'none', () => {
    if (typeof vote !== 'undefined') {
      if (vote.vote_choice === 1) {
        setPoints(pointState + 1);
        return 'upvote';
      }
      if (vote.vote_choice === -1) {
        setPoints(pointState - 1);
        return 'downvote';
      }
    }
    return 'none';
  });

  return (
    <>
      <CardContainerStyle>
        <TitleStyle style={{ gridColumn: 1, gridRow: 1 }}>{title}</TitleStyle>
        <div style={{ gridColumn: '1 / span 2', gridRow: 2 }}>
          <ItemStyle>{subject}</ItemStyle>
          <ItemStyle>{topic}</ItemStyle>
          <div style={{ height: '5px' }} />
        </div>
        <AvatarStyle
          style={{ gridColumn: 2, gridRow: '1 / span 2', borderRadius: '50%' }}
          width="40px"
          height="40px"
          src={authorImg}
          alt="uploader avatar"
        />
        <ItemStyle style={{ gridColumn: '1 / span 2', gridRow: 3 }}>Uploaded by {author}</ItemStyle>
        <PreviewA href={url}>
          <PreviewStyle width="100%" height="195px" src={previewImg} alt="preview" />
        </PreviewA>
        <PreviewLink href={url}>
          <LinkStyle>{url.length < 37 ? url : `${url.substring(0, 34)}...`}</LinkStyle>
        </PreviewLink>
        <ItemStyle
          style={{
            color: `${(() => {
              if (voteState === 'upvote') {
                return '#ff945e';
              }
              if (voteState === 'downvote') {
                return '#7fdbff';
              }
              return '';
            })()}`,
            gridColumn: '1 / span 2',
            gridRow: 6
          }}
        >
          {pointState > 0 ? `+${pointState}` : `${pointState}`} points
        </ItemStyle>
        <ItemStyle style={{ marginBottom: '6px', gridColumn: '1 / span 2', gridRow: 7 }}>
          Uploaded {dateToDisplay}
        </ItemStyle>
        <DescriptionStyle>{description === '' ? 'no description provided' : description}</DescriptionStyle>
        <BottomContainer>
          {signedInAs === '' ? (
            <a href="/signIn">Sign in to vote!</a>
          ) : (
            <>
              <UnstyledButton data-testid={`${title}-upvote-arrow`} onClick={() => dispatch('upvote')}>
                <FontAwesomeIcon color={voteState === 'upvote' ? '#ff945e' : '#111111'} icon="arrow-up" />
              </UnstyledButton>
              <UnstyledButton data-testid={`${title}-downvote-arrow`} onClick={() => dispatch('downvote')}>
                <FontAwesomeIcon color={voteState === 'downvote' ? '#7fdbff' : '#111111'} icon="arrow-down" />
              </UnstyledButton>
              <UnstyledButton data-testid={`${title}-options`} onClick={() => setShowingDropDown(!isShowingDropDown)}>
                <FontAwesomeIcon color="#111111" icon="ellipsis-v" />
              </UnstyledButton>
            </>
          )}
        </BottomContainer>
        {isShowingDropDown && (
          <ResourceCardDropDown>
            <UnstyledButton
              style={{
                color: '#0074d9',
                backgroundColor: '#efefef',
                width: '100%',
                height: '32px',
                borderBottom: '1px solid #e3e3e3',
                display: 'flex',
                alignItems: 'center',
                cursor: 'pointer'
              }}
              onClick={() => setShowingReportModal(true)}
            >
              Report...
            </UnstyledButton>
            <UnstyledButton
              style={{
                color: '#0074d9',
                backgroundColor: '#efefef',
                width: '100%',
                height: '32px',
                borderBottom: '1px solid #e3e3e3',
                display: 'flex',
                alignItems: 'center',
                cursor: 'pointer'
              }}
              onClick={() => setShowingImageModal(true)}
            >
              Add Image
            </UnstyledButton>
            {userData.is_staff === true && (
              <UnstyledButton
                style={{
                  color: '#ff4300',
                  backgroundColor: '#efefef',
                  width: '100%',
                  height: '32px',
                  borderBottom: '1px solid #e3e3e3',
                  display: 'flex',
                  alignItems: 'center',
                  cursor: 'pointer'
                }}
                onClick={() => {
                  const token = localStorage.getItem('token');
                  axios.post(
                    `${urlToUse}/v1/post/deletePost/${id}/`,
                    {},
                    {
                      headers: { Authorization: `Bearer ${token}` }
                    }
                  );
                }}
              >
                Delete
              </UnstyledButton>
            )}
          </ResourceCardDropDown>
        )}
      </CardContainerStyle>
      <Modal isOpen={isShowingReportModal}>
        <div style={{ width: '100%', display: 'grid', gridTemplateColumns: 'auto 36px', cursor: 'pointer' }}>
          <FontAwesomeIcon
            style={{ gridRow: 1, gridColumn: 2 }}
            onClick={() => setShowingReportModal(false)}
            icon="times"
          />
          <h1 style={{ color: '#0074d9' }}>Report Post: {title}</h1>
          <Formik
            initialValues={{ description: '' }}
            onSubmit={reportPost}
            render={formikProps => {
              const { handleChange, handleSubmit, values } = formikProps;
              return (
                <form onSubmit={handleSubmit}>
                  <TextAreaStyle
                    onChange={handleChange}
                    name="description"
                    value={values.description}
                    placeholder="What's wrong with this post?"
                    style={{ gridColumn: '1 / span 2' }}
                  />
                  <GenericButton type="submit" text="SUBMIT" />
                </form>
              );
            }}
          />
        </div>
      </Modal>
      <Modal onAfterOpen={() => setShowingReportModal(false)} isOpen={isShowingImageModal}>
        <div style={{ width: '100%', display: 'grid', gridTemplateColumns: 'auto 36px', cursor: 'pointer' }}>
          <FontAwesomeIcon
            style={{ gridRow: 1, gridColumn: 2 }}
            onClick={() => setShowingImageModal(false)}
            icon="times"
          />
          <h1 style={{ color: '#0074d9' }}>Add Image to Post: {title}</h1>
          <p style={{ color: '#0074d9', gridColumn: '1 / span 2' }}>Image must be a url</p>
          <Formik
            initialValues={{ url: '' }}
            onSubmit={addImageToPost}
            render={formikProps => {
              const { handleChange, handleSubmit, values } = formikProps;
              return (
                <form onSubmit={handleSubmit}>
                  <TextAreaStyle
                    onChange={handleChange}
                    name="url"
                    type="url"
                    placeholder="url"
                    value={values.url}
                    style={{ gridColumn: '1 / span 2' }}
                  />
                  <GenericButton type="submit" text="SUBMIT" />
                </form>
              );
            }}
          />
        </div>
      </Modal>
    </>
  );
};

ResourceCard.propTypes = {
  title: PropTypes.string.isRequired,
  subject: PropTypes.string.isRequired,
  topic: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  authorImg: PropTypes.string.isRequired,
  previewImg: PropTypes.string.isRequired,
  points: PropTypes.number.isRequired,
  timeStamp: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  vote: PropTypes.shape({})
};

ResourceCard.defaultProps = {
  vote: undefined
};

export default ResourceCard;
