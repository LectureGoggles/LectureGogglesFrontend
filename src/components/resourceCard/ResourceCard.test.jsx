import React from 'react';
import { render, cleanup } from 'react-testing-library';
import ResourceCard from './ResourceCard';

afterEach(cleanup);

/**
 * Tests for text rendering
 */
it('Renders a title', () => {
  const { getByText } = render(
    <ResourceCard
      title="A Title I Have Chosen"
      subject=""
      topic=""
      author=""
      authorImg=""
      previewImg=""
      points={0}
      timeStamp=""
      description=""
      url=""
    />
  );

  expect(getByText(/A Title I Have Chosen/gm)).toBeTruthy();
  expect(getByText(/A Title I Have Chosen/gm).outerHTML).toMatch(/<div class=".*">A Title I Have Chosen<\/div>/gm);
});

// it('Renders a Subject', () => {
//   const { getByText } = render(
//     <ResourceCard
//       title=""
//       subject="Some kind of subject"
//       topic=""
//       author=""
//       authorImg=""
//       previewImg=""
//       points={0}
//       timeStamp=""
//       description=""
//       url=""
//     />
//   );

//   expect(getByText(/Some kind of subject/gm)).toBeTruthy();
//   expect(getByText(/Some kind of subject/gm).outerHTML).toMatch(/<div class=".*">Some kind of subject - <\/div>/gm);
// });

// it('Renders a topic', () => {
//   const { getByText } = render(
//     <ResourceCard
//       title=""
//       subject=""
//       topic="Some random topic"
//       author=""
//       authorImg=""
//       previewImg=""
//       points={0}
//       timeStamp=""
//       description=""
//       url=""
//     />
//   );

//   expect(getByText(/Some random topic/gm)).toBeTruthy();
//   expect(getByText(/Some random topic/gm).outerHTML).toMatch(/<div class=".*"> - Some random topic<\/div>/gm);
// });

it('Renders an author', () => {
  const { getByText } = render(
    <ResourceCard
      title=""
      subject=""
      topic=""
      author="Jane Doe"
      authorImg=""
      previewImg=""
      points={0}
      timeStamp=""
      description=""
      url=""
    />
  );

  expect(getByText(/Jane Doe/gm)).toBeTruthy();
  expect(getByText(/Jane Doe/gm).outerHTML).toMatch(/<div class=".*">Uploaded by Jane Doe<\/div>/gm);
});

it('Renders an image of the author with an alt image text', () => {
  const { getByAltText } = render(
    <ResourceCard
      title=""
      subject=""
      topic=""
      author=""
      authorImg="some_url_would_be_here"
      previewImg=""
      points={0}
      timeStamp=""
      description=""
      url=""
    />
  );

  expect(getByAltText(/uploader avatar/gm)).toBeTruthy();
  expect(getByAltText(/uploader avatar/gm).outerHTML).toMatch(
    /<img width=".*px" height=".*px" src="some_url_would_be_here" alt="uploader avatar" class=".*"/gm
  );
});

it('Renders an image of the author with an alt image text', () => {
  const { getByAltText } = render(
    <ResourceCard
      title=""
      subject=""
      topic=""
      author=""
      authorImg=""
      previewImg="a_url_goes_here"
      points={0}
      timeStamp=""
      description=""
      url="a_link"
    />
  );

  expect(getByAltText(/preview/gm)).toBeTruthy();
  expect(getByAltText(/preview/gm).outerHTML).toMatch(
    /<img width=".*" height=".*" src="a_url_goes_here" alt="preview" class=".*">/gm
  );
});

it('Includes a hyperlink of the url on the image', () => {
  const { getByAltText } = render(
    <ResourceCard
      title=""
      subject=""
      topic=""
      author=""
      authorImg=""
      previewImg=""
      points={0}
      timeStamp=""
      description=""
      url="a_link"
    />
  );

  expect(getByAltText(/preview/gm)).toBeTruthy();
  expect(getByAltText(/preview/gm).parentElement.outerHTML).toMatch(
    /<a href="a_link" class=".*"><img width=".*" height=".*" src="" alt="preview" class=".*"> <\/a>/gm
  );
});

it('Renders the url of the resource', () => {
  const { getByText } = render(
    <ResourceCard
      title=""
      subject=""
      topic=""
      author=""
      authorImg=""
      previewImg=""
      points={0}
      timeStamp=""
      description=""
      url="a_url"
    />
  );

  expect(getByText(/a_url/gm)).toBeTruthy();
  expect(getByText(/a_url/gm).outerHTML).toMatch(/<div class=".*">a_url<\/div>/gm);
});

it('Truncates a long url of the resource', () => {
  const url =
    'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa';
  const { getByText } = render(
    <ResourceCard
      title=""
      subject=""
      topic=""
      author=""
      authorImg=""
      previewImg=""
      points={0}
      timeStamp=""
      description=""
      url={url}
    />
  );

  expect(getByText(/a{34}.../gm)).toBeTruthy();
});

it('Renders the points of a post', () => {
  const { getByText } = render(
    <ResourceCard
      title=""
      subject=""
      topic=""
      author=""
      authorImg=""
      previewImg=""
      points={-2}
      timeStamp=""
      description=""
      url=""
    />
  );

  expect(getByText(/-2/gm)).toBeTruthy();
  expect(getByText(/-2/gm).outerHTML).toMatch(/<div.*>-2 points<\/div>/gm);
});

it('Renders the positive points of a post', () => {
  const { getByText } = render(
    <ResourceCard
      title=""
      subject=""
      topic=""
      author=""
      authorImg=""
      previewImg=""
      points={20}
      timeStamp=""
      description=""
      url=""
    />
  );

  expect(getByText(/\+20/gm)).toBeTruthy();
  expect(getByText(/\+20/gm).outerHTML).toMatch(/<div .*>\+20 points<\/div>/gm);
});

// Timestamp tests will depend on how the api and database format the data. Test tbd.

it('Renders a description of a post', () => {
  const des = 'Brief description of some post';
  const { getByText } = render(
    <ResourceCard
      title=""
      subject=""
      topic=""
      author=""
      authorImg=""
      previewImg=""
      points={0}
      timeStamp=""
      description={des}
      url=""
    />
  );
  expect(getByText(/Brief description of some post/gm)).toBeTruthy();
  expect(getByText(/Brief description of some post/gm).outerHTML).toMatch(
    /<div class=".*">Brief description of some post<\/div>/gm
  );
});

// For integration
it('Sends an upvote when upvote button is pressed', () => {});
it('Sends an downvote when upvote button is pressed', () => {});

// These tests will change when the resource card is finished
it('Renders an upvote button', () => {
  const { queryByTestId } = render(
    <ResourceCard
      title="resource"
      subject=""
      topic=""
      author=""
      authorImg=""
      previewImg=""
      points={0}
      timeStamp=""
      description=""
      url=""
    />
  );

  expect(queryByTestId('resource-upvote-arrow').outerHTML).toMatch(
    /<button data-testid="resource-upvote-arrow" class=".*"><i class="fa"><\/i><\/button>/gm
  );
});

it('Renders a downvote button', () => {
  const { queryByTestId } = render(
    <ResourceCard
      title="resource"
      subject=""
      topic=""
      author=""
      authorImg=""
      previewImg=""
      points={0}
      timeStamp=""
      description=""
      url=""
    />
  );

  expect(queryByTestId('resource-downvote-arrow').outerHTML).toMatch(
    /<button data-testid="resource-downvote-arrow" class=".*"><i class="fa"><\/i><\/button>/gm
  );
});
