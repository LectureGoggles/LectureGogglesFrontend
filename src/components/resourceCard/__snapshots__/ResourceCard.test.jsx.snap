// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches the snapshot 1`] = `"<div class=\\"sc-bdVaJa jBUENB\\"><div class=\\"sc-bwzfXH iprmFX\\">a</div><div class=\\"sc-htpNat gjdmzU\\">b - c</div><div class=\\"sc-htpNat gjdmzU\\">Uploaded by d</div><img width=\\"40px\\" height=\\"40px\\" src=\\"e\\" alt=\\"uploader avatar\\" class=\\"sc-ifAKCX fKTzex\\"><a href=\\"i\\" class=\\"sc-bZQynM eYttIo\\"><img width=\\"100%\\" height=\\"195px\\" src=\\"f\\" alt=\\"preview\\" class=\\"sc-EHOje gckEHp\\"> </a><div class=\\"sc-gzVnrw gPooSX\\"><div class=\\"sc-htoDjs ijqXLH\\">i</div></div><div class=\\"sc-htpNat gjdmzU\\">+123 points</div><br><div class=\\"sc-htpNat gjdmzU\\">Uploaded g ago</div><br><div class=\\"sc-htpNat sc-bxivhb lePNbT\\">h</div><br><div class=\\"sc-bwzfXH iprmFX\\"><button type=\\"button\\">↑ Upvote</button><button type=\\"button\\">↓ Downvote</button></div></div>"`;
